### ISSUE TITLE
Unexpected return type from equals. Expected Nullable(UInt8). Got UInt8. In old analyzer

### ISSUE THREAD

**Issue by Avogar** [ClickHouse/ClickHouse#66829]
How to reproduce:
```sql
CREATE TABLE test
(
    `test1_id` UUID,
    `test2_id` UUID
)
ENGINE = MergeTree
ORDER BY (test1_id, test2_id)
SETTINGS index_granularity = 8192;

CREATE TABLE test1
(
    `id` UUID,
    `test2_id` UUID
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192;

CREATE TABLE test2
(
    `id` UUID,
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192;

WITH CAST('00000000-0000-0000-0000-000000000000', 'Nullable(UUID)') AS test1_id
SELECT *
FROM test AS t
INNER JOIN test1 AS t1 ON t.test1_id = t1.id
INNER JOIN test2 AS t2 ON t1.test2_id = t2.id
WHERE test1.id = test1_id;
```

```
Received exception from server (version 24.7.1):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Unexpected return type from equals. Expected Nullable(UInt8). Got UInt8. Action:
FUNCTION equals(id : 1, _--t1.id : 0) -> equals(id, _--t1.id) Nullable(UInt8) : 3,
input block structure:id UUID UUID(size = 0), _--t1.id UUID UUID(size = 0): while executing 'FUNCTION equals(id : 1, _--t1.id : 0) -> equals(id, _--t1.id) Nullable(UInt8) : 3'. (LOGICAL_ERROR)
```

It works with `allow_epxerimental_analyzer=0` or with `query_plan_filter_push_down = 0`.

**Comment by yakov-olkhovskiy** [comment]
will work on it...

**Comment by yakov-olkhovskiy** [comment]
TL/DR
1. this bug is not curable without addressing more fundamental issues of aliasing and names' collision;
2. workaround - `set prefer_column_name_to_alias = 1`;
3. since we are going to push forward the new analyzer and workaround exists for this one probably we can left it as is.

@alexey-milovidov @Avogar your thoughts?

Explanation:
I have found that aliasing in general doesn't work consistently - for example:
```
set allow_experimental_analyzer=0;

set prefer_column_name_to_alias = 0;

CREATE TABLE test
(
    `test1_id` UUID,
    `test2_id` UUID
)
ENGINE = MergeTree
ORDER BY (test1_id, test2_id)
SETTINGS index_granularity = 8192;

EXPLAIN PLAN actions=1,header=1
WITH CAST('00000000-0000-0000-0000-000000000000', 'Nullable(UUID)') AS test1_id
SELECT *
FROM test as t
WHERE test2_id = test1_id;

Expression ((Projection + Before ORDER BY))
Header: test1_id UUID
        test2_id UUID
Actions: INPUT :: 0 -> test1_id UUID : 0
         INPUT :: 1 -> test2_id UUID : 1
Positions: 0 1
  Filter (WHERE)
  Header: test1_id UUID
          test2_id UUID
  Filter column: equals(test2_id, test1_id) (removed)
  Actions: INPUT : 0 -> test1_id UUID : 0
           INPUT : 1 -> test2_id UUID : 1
           FUNCTION equals(test2_id : 1, test1_id : 0) -> equals(test2_id, test1_id) UInt8 : 2
  Positions: 0 1 2
    ReadFromPreparedSource (Read from NullSource)
    Header: test1_id UUID
            test2_id UUID
```
In this case `test1_id` field of `WHERE` clause should be chosen depending on `prefer_column_name_to_alias` setting - if disabled then aliased value from `WITH` should be taken - it is not happening in this case, but it is happening in (some) cases with `JOIN` - like one in the example of this issue. The consequent error is produced by optimization of `WHERE` clause which is pushed down to joined tables - at this stage it fails to recognize aliased field and treats it as table's column with corresponding name while equal function's return type `Nullable(UUID)` left intact, but since now both arguments are not nullable the actual return type is not nullable (the rule for `equal` function is that when any argument is nullable - result is nullable, if non is nullable - result is not nullable) - thus the error.

**Comment by alexey-milovidov** [comment]
Yes, it is ok to avoid fixing this bug, but we need to somehow mute it in the CI.

**Comment by yakov-olkhovskiy** [comment]
we don't have this in CI - the issue was caught a couple times in production: https://github.com/ClickHouse/clickhouse-core-incidents/issues/94
I see another problem though - if for the time being this issue will surface in production again we need a way for person working with customer to recall this decision and to recommend appropriate workaround - do we have a centralized catalog of such things? it's quite awkward to search issues and PRs in hope to find it...

**Comment by yakov-olkhovskiy** [comment]
### Proposed workarounds in order of priority:
0. Change aliasing to avoid collisions.
1. Switch to new analyzer:
```
set allow_experimental_analyzer = 1;
```
2. Restrict alias propagation:
```
set prefer_column_name_to_alias = 1;
```
3. Disable filter push down optimization:
```
set query_plan_filter_push_down = 0;
```
----------------------------
Closing this issue for now.

**Comment by al13n321** [comment]
Another repro with analyzer and `join_use_nulls`, still based on name collision:
```sql
WITH
    a AS (SELECT 42::UInt64 as x),
    b AS (SELECT 0::UInt64 as x, '' as y)
SELECT
    countIf(b.y = '') AS b
FROM a
LEFT JOIN b ON b.x = a.x
SETTINGS join_use_nulls = 1, allow_experimental_analyzer=1
```
